src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+using ArgCheck
+using OffsetArrays
+
+export tmap, tmap!, tmapreduce, treduce
+
+############################## docs ##############################
+function _make_docstring(fname, args, kw=String[])
+    tf = tname(Symbol(fname))
+    sargs = join(args, ", ")
+    skw = join([kw; "[batch_size::Integer]"], ", ")
+    signature = string(tf, "(", sargs, "; ", skw, ")")
+    """
+        $signature
+
+    Threaded analog of [`Base.$fname`](@ref). See [`Base.$fname`](@ref) for a description of arguments.
+    In order to parallelize the arguments are split into chunks whose size in controlled by the `batch_size` keyword.
+    """
+end
+
+function tname(s::Symbol)
+    Symbol("t", s)
+end
+
+const SYMBOLS_MAPREDUCE_LIKE = [:sum, :prod, :minimum, :maximum]
+for fun in SYMBOLS_MAPREDUCE_LIKE
+    tfun = tname(fun)
+    docstring = _make_docstring(fun, ["[f]", "src::AbstractArray"])
+    @eval begin
+        export $tfun
+        @doc $docstring ->
+        function $tfun end
+    end
+end
+
+@doc _make_docstring(:map!, ["f", "dst::AbstractArray", "srcs::AbstractArray..."]) ->
+function tmap! end
+
+@doc _make_docstring(:map, ["f", "srcs::AbstractArray..."]) ->
+function tmap end
+
+@doc _make_docstring(:mapreduce, ["f", "op", "src::AbstractArray"], ["[init]"]) ->
+function tmapreduce end
+
+@doc _make_docstring(:reduce, ["op", "src::AbstractArray"],  ["[init]"]) ->
+function treduce end
+
+############################## Helper functions ##############################
+struct Batches{V}
+    firstindex::Int
+    lastindex::Int
+    batch_size::Int
+    values::V
+    length::Int
+end
+
+function Batches(values, batch_size::Integer)
+    @argcheck batch_size > 0
+    r = eachindex(values)
+    @assert r isa AbstractUnitRange
+    len = ceil(Int, length(r) / batch_size)
+    Batches(first(r), last(r), batch_size, values, len)
+end
+
+Base.length(o::Batches) = o.length
+Base.eachindex(o::Batches) = Base.OneTo(length(o))
+function Base.getindex(o::Batches, i)
+    @boundscheck (@argcheck i in eachindex(o))
+    start = o.firstindex + (i-1) * o.batch_size
+    stop  = min(start + (o.batch_size) -1, o.lastindex)
+    o.values[start:stop]
+end
+
+function default_batch_size(len)
+    len <= 1 && return 1
+    nthreads=Threads.nthreads()
+    items_per_thread = len/nthreads
+    items_per_batch = items_per_thread/4
+    clamp(1, round(Int, items_per_batch), len)
+end
+
+function Base.iterate(o::Batches, state=1)
+    if state in eachindex(o)
+        o[state], state+1
+    else
+        nothing
+    end
+end
+
+mutable struct _RollingCutOut{A,I<:AbstractUnitRange,T} <: AbstractVector{T}
+    array::A
+    eachindex::I
+end
+
+function _RollingCutOut(array::AbstractArray, indices)
+    T = eltype(array)
+    A = typeof(array)
+    I = typeof(indices)
+    _RollingCutOut{A, I, T}(array, indices)
+end
+
+Base.size(r::_RollingCutOut) = (length(r.eachindex),)
+
+function Base.eachindex(r::_RollingCutOut, rs::_RollingCutOut...)
+    for r2 in rs
```

Comment:
Potentially type unstable.

Reply:
When the `rs` elements have different types.

Reply:
You really have an eye for type instability, I like that. I will fix it (though again I think in practice it won't matter.)

Reply:
Thanks! It won't matter until it does! If this is part of a loop, then the overhead will add up.